### PR TITLE
[SYCL] Support standard library functions for AOT compilation

### PR DESCRIPTION
**Pull request by asavonic**
Fallback libraries are now compiled into object files, so they can be
linked at compile time.

Signed-off-by: Andrew Savonichev <andrew.savonichev@intel.com>

### CODE REVIEW

File: sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst [hunk 1]
```diff
@@ -270,6 +280,6 @@ Every set of functions is implemented in a separate fallback
 library. For example, a fallback for `cl_intel_devicelib_cassert`
 extension is provided as `libsycl-fallback-cassert.spv`
 
-NOTE that AOT compilation is not yet supported. Driver will have to
-check for extension support and link the corresponding SPIR-V fallback
-implementation, but this is not implemented yet.
+For AOT compilation, fallback libraries are also provided as object
+files (e.g. `libsycl-fallback-cassert.o`). Device code in these object
+files is identical to the the `*.spv` files.
```
Comment (kbobrovs) [inline sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:285]: I'd expect the device code in .o files to be llvm IR bitcode, if it is produced with '-fsycl -c'.
Is this really a SPIR-V now?

Comment (asavonic) [reply sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst:285]: @vzakhari is correct. I will reword this section.

File: libdevice/cmake/modules/SYCLLibdevice.cmake [hunk 2]
```diff
@@ -125,6 +168,15 @@ add_custom_command(OUTPUT ${binary_dir}/libsycl-fallback-cmath-fp64.spv
                    DEPENDS device_math.h device.h clang llvm-spirv
                    VERBATIM)
 
+add_custom_command(OUTPUT ${binary_dir}/libsycl-fallback-cmath-fp64.o
+                   COMMAND ${clang} -fsycl -c
+                           ${compile_opts} ${sycl_targets_opt}
+                           ${CMAKE_CURRENT_SOURCE_DIR}/fallback-cmath-fp64.cpp
+                           -o ${binary_dir}/libsycl-fallback-cmath-fp64.o
+                   MAIN_DEPENDENCY fallback-cmath-fp64.cpp
+                   DEPENDS device_math.h device.h clang
```
[the remark this answers is not in the document]
Comment (asavonic) [reply libdevice/cmake/modules/SYCLLibdevice.cmake]: Done.

File: sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst (hunk 1 above)
Comment (asavonic) [reply sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst]: Done. Please check if it is OK now.

File: libdevice/cmake/modules/SYCLLibdevice.cmake [hunk 3]
```diff
@@ -89,13 +96,31 @@ add_custom_command(OUTPUT ${binary_dir}/libsycl-fallback-cassert.spv
...
+                   VERBATIM)
+
 add_custom_command(OUTPUT ${binary_dir}/libsycl-fallback-complex.spv
                    COMMAND ${clang} -S -fsycl-device-only -fno-sycl-use-bitcode
                            ${compile_opts}
                            ${CMAKE_CURRENT_SOURCE_DIR}/fallback-complex.cpp
                            -o ${binary_dir}/libsycl-fallback-complex.spv
                    MAIN_DEPENDENCY fallback-complex.cpp
-                   DEPENDS device_math.h device_complex.h device.h clang llvm-spirv
+                   DEPENDS device_math.h device_complex.h device.h clang llvm-spirv clang-offload-bundler
```
Comment (vzakhari) [inline libdevice/cmake/modules/SYCLLibdevice.cmake:114]: nit: clang-offload-bundler is not needed here.

Comment (asavonic) [reply libdevice/cmake/modules/SYCLLibdevice.cmake]: Thanks, fixed.

File: sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst (hunk 1 above)
Comment (kbobrovs) [reply sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst]: LGTM, thanks

**Pull request merged**